client/state/ui/layout-focus/reducer.js

```diff
@@ -0,0 +1,42 @@
+/**
+ * External dependencies
+ */
+import includes from 'lodash/includes';
+
+/**
+ * Internal dependencies
+ */
+import {
+	LAYOUT_FOCUS_SET,
+	LAYOUT_NEXT_FOCUS_ACTIVATE,
+	LAYOUT_NEXT_FOCUS_SET,
+} from 'state/action-types';
+
+const initialState = { current: null, previous: null, next: null };
+const validAreas = [ 'content', 'sidebar', 'sites', 'preview' ];
+const isValidArea = area => includes( validAreas, area );
+
+export default function layoutFocus( state = initialState, action ) {
+	switch ( action.type ) {
+		case LAYOUT_FOCUS_SET:
+			if ( action.area === state.current || ! isValidArea( action.area ) ) {
+				return state;
+			}
+			return Object.assign( {}, state, { current: action.area, previous: state.current } );
+		case LAYOUT_NEXT_FOCUS_SET:
+			if ( action.area === state.next || ! isValidArea( action.area ) ) {
+				return state;
+			}
+			return Object.assign( {}, state, { next: action.area } );
+		case LAYOUT_NEXT_FOCUS_ACTIVATE:
+			// If we don't have a change queued and the focus has changed
+			// previously, set it to `content`. This avoids having to set the
+			// focus to content on all navigation links because it becomes the
+			// default after focus has shifted.
+			if ( ! state.next && ! state.previous ) {
```

Comment:
Hm... I thought I grasped it, but I'm still not sure. I re-wrote that part of the reducer so that the logic is the same, but is now in the style of the original module. I guess what I'm not clear on is: when exactly is `activateNextLayoutFocus` called? It seems to be only in that weird `loader` file, but I don't understand when that happens in the normal process of navigation.